component-base: test and fix JSON output for KObj

pohly opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

KObj wasn't get logged as struct when using JSON output.

#### Does this PR introduce a user-facing change?
```release-note
Kubernetes object references (= name + namespace) were not logged as struct when using JSON as log output format.
```

/cc @serathius @thockin

pohly committed:
component-base: test json output for KObj and KObjs

Thanks to the new klog, logr and zapr releases, ObjectRef is now
logged as struct in JSON output.

serathius commented on staging/src/k8s.io/component-base/logs/json/klog_test.go:
```diff
@@ -116,6 +118,13 @@ func TestKlogIntegration(t *testing.T) {
 			format: `{"ts":%f,"caller":"json/klog_test.go:%d","msg":"test","v":0,"count":1}`,
 		},
+		{
+			name: "KObj",
```
Can you also add a test for KObjs?

pohly replied:
Not yet - https://github.com/kubernetes/kubernetes/pull/104103 needs to be merged first.

But I've tested locally based on that PR and KObjs prints the slice as structs:
```
		{
			name: "KObjs",
			fun: func() {
				klog.InfoS("several", "pods",
					klog.KObjs([]interface{}{
						&kmeta{Name: "pod-1", Namespace: "kube-system"},
						&kmeta{Name: "pod-2", Namespace: "kube-system"},
					}))
			},
			format: `{"ts":%f,"caller":"json/klog_test.go:%d","msg":"several","v":0,"pods":[{"name":"pod-1","namespace":"kube-system"},{"name":"pod-2","namespace":"kube-system"}]}`,
		},
```

serathius replied:
heh, that makes sense as slice doesn't implement Stringer.

serathius commented:
Sad that we missed this, I think I remember noticing this at some point, but when final implementation came we didn't have tests. We should definetly fix it for 1.23, @pohly can you file a issue to address this. We should track it in the Structured Logging KEP.

pohly commented:
/hold

Let's use this PR to include a fix, if there is one.

fedebongio commented:
/assign @serathius 
/triage accepted

serathius commented:
/lgtm
/approve

serathius commented:
/retest

pohly commented:
/kind bug

pohly committed:
klog 2.30.0, logr 1.2.0, zapr 1.2.0

The new releases fix logging of KObj in JSON output: klog implements the new
logr.Marshaler interface and zapr uses it instead of Stringer when logging the
ObjectRef created by KObj.
